### ISSUE TITLE
pyDat 4.0 status - what branch to use today

### ISSUE THREAD

**Issue by abgoldberg** [MITRECND/WhoDat#16]
Thanks for the useful project. 

I recently purchased whoisxmlapi data and am looking into pyDat for running my local API on top of the raw data. I was wondering whether I should start using the current release (master branch) or if one of the other branches with support for newer ES versions is production-ready (or close enough).

Thanks in advance. (If there's a better way to communicate about general questions about the project other than an Issue, just let me know.)

**Comment by Mraoul** [comment]
The pydat4-dev branch is your best bet, but at best, I'd call it 'beta' level. Proper support for ES 6.x was added very recently (like last week) -- but the current head of that branch should work just fine with ES 5.2 through to ES 6.3. The biggest issue with this branch is if any changes need to be made to the ElasticSearch template, it would require reingestion of the data, which is extremely time consuming, especially if you have multiple versions of the whois data loaded. So, if you do decide to use the pydat4-dev branch, consider that you might have to reingest your data at some point if the template changes in a significant fashion.

If you do decide to use the pydat4-dev branch, any feedback, issues, etc would be appreciated.

**Comment by abgoldberg** [comment]
Thanks! I will try the pydat4-dev branch. Will that work via the current docker instructions, or is my best bet to install it all manually?

**Comment by Mraoul** [comment]
I just created a hook on docker hub to auto-build the pydat4-dev branch, give it ~20 minutes and you should be able to pull `mitrecnd/pydat:pydat4-dev`

**Comment by abgoldberg** [comment]
I've run into a few issues getting things going...

I ran the pydat4-dev docker container and at first got this log output:

`AH00534: httpd: Configuration error: No MPM loaded.`

I think I managed to get around that by tweaking the Dockerfile to have this:

```
(echo "LoadModule mpm_event_module modules/mod_mpm_event.so"; cat /opt/WhoDat/docker/apache.config) > /opt/WhoDat/docker/apache.config.mod && \
cp /opt/WhoDat/docker/apache.config.mod /usr/local/apache2/conf/httpd.conf && \
```

and rebuilt the container. So httpd can start up now. 

But with the server apparently running, I only get responses of status 400 Bad Request. The log has these permissions errors:

```
AH00558: httpd: Could not reliably determine the server's fully qualified domain name, using 172.17.0.2. Set the 'ServerName' directive globally to suppress this message
AH00558: httpd: Could not reliably determine the server's fully qualified domain name, using 172.17.0.2. Set the 'ServerName' directive globally to suppress this message
[Fri Jul 13 16:17:53.996323 2018] [mpm_event:notice] [pid 1:tid 140054308788096] AH00489: Apache/2.4.33 (Unix) mod_wsgi/4.4.13 Python/2.7.9 configured -- resuming normal operations
[Fri Jul 13 16:17:53.996741 2018] [core:notice] [pid 1:tid 140054308788096] AH00094: Command line: 'httpd -D FOREGROUND'
[Fri Jul 13 11:20:54.453756 2018] [wsgi:error] [pid 7:tid 140054163273472] WARNING: Couldn't open 'parser.out'. [Errno 13] Permission denied: '/opt/WhoDat/pydat/pydat/handlers/parser.out'
[Fri Jul 13 11:20:54.458124 2018] [wsgi:error] [pid 7:tid 140054163273472] Generating LALR tables
[Fri Jul 13 11:20:54.477576 2018] [wsgi:error] [pid 7:tid 140054163273472] WARNING: Couldn't create 'handlers.parsetab'. [Errno 13] Permission denied: '/opt/WhoDat/pydat/pydat/handlers/parsetab.py'
[Fri Jul 13 11:20:54.477979 2018] [wsgi:error] [pid 7:tid 140054163273472] WARNING: No PDNS modules are active. PyDat's Passive DNS functionality can not be conducted without any activated PDNS modules.
[Fri Jul 13 11:20:54.477998 2018] [wsgi:error] [pid 7:tid 140054163273472] 
[Fri Jul 13 11:20:54.478186 2018] [wsgi:error] [pid 7:tid 140054163273472] WARNING: No PDNS modules are active. PyDat's Passive DNS functionality can not be conducted without any activated PDNS modules.
[Fri Jul 13 11:20:54.478201 2018] [wsgi:error] [pid 7:tid 140054163273472] 
172.17.0.1 - - [13/Jul/2018:16:20:53 +0000] "GET /ajax/domain/google.org/latest/ HTTP/1.1" 400 26
```
 I'm new to docker so I'm not sure what user is executing at that point and the best way to go about fixing that. Any suggestions?

**Comment by Mraoul** [comment]
I'll have to fix the first error.

For the second error -- how do you have ElasticSearch setup and connected to pydat? 

The parser error _shouldn't_ be a critical error -- basically the grammar parser compiles it and writes it to disk so that it doesn't need to recompile it on the next run, but an existing process keeps it in memory so it's not a huge deal for a long-running process.

**Comment by abgoldberg** [comment]
Thanks. I see now the result is the same even if i add a line to Dockerfile to change the permissions on the /opt/WhoDat directory.

I have ElasticSearch running via docker like so:

`sudo docker run -d -p 9200:9200 -p 9300:9300 -e "discovery.type=single-node" docker.elastic.co/elasticsearch/elasticsearch:6.3.1`

I then (seemingly successfully) ran this to populate it:

`python elasticsearch_populate.py -u localhost:9200 -f ~/whois_data/v24/csv/tlds/simple/simple/org/1.csv -i '1' -v -s`

I'm running the pydat container with `-p 12345:80` and attempting to query it using:

`curl http://localhost:12345/ajax/domain/google.org/latest/`

**Comment by Mraoul** [comment]
Try the regular web endpoint and see if you get an error. I tried a similar test and got back a JSON response, so I'm guessing it might be an ES connectivity error -- although, that usually throws back a different kind of error.

Can you paste your custom_settings.py file (only the ES-related stuff)? And what command you used to start up the pydat container?

One thing to note about docker container connectivity is separate containers generally have to be linked together via something like a docker network (or the legacy --link option). Also, are you familiar with docker-compose, this might be easier to setup/test with compose.

**Comment by abgoldberg** [comment]
Thanks a lot for helping debug this.

Same result on port 80. I didn't actually change anything from the default
ES settings.

The command to run the container is:

`sudo docker run --name pydat-custom -p 80:80 -v
~/WhoDat/pydat/pydat/custom_settings.py:/opt/WhoDat/pydat/pydat/custom_settings.py
79a4de197458`
(note i'm using a custom name and have my own container ID since i rebuilt
the container with the Dockerfile changes mentioned above)

You're probably on to something with the docker container connectivity. I
am not familiar with docker-compose at all but I can look into that, or
this linking option you mentioned.

**Comment by Mraoul** [comment]
I've updated the repo with some instructions on how to deploy it via docker-compose, not sure if that'll help or not. Also, I pushed a patch to the apache config so the latest version of pydat4-dev on docker hub should run properly.

**Comment by abgoldberg** [comment]
The new docker-compose approach works. When I first ran it, I got this
error and elasticsearch quitting:

`elasticsearch_1  | [1]: max virtual memory areas vm.max_map_count [65530]
is too low, increase to at least [262144]`

but running

`sudo sysctl -w vm.max_map_count=262144`

and trying again got past that.

Now it seems like when I load a lot of data into elasticsearch, it exits
with code 137, which appears to be OOM. I'll experiment with different
memory limit parameters. Or maybe just consider running with a very small
subset a big enough victory for now, and then move on to a more
production-like setup with a EC cluster.

**Comment by Mraoul** [comment]
Yeah, trying to load the full set with the memory settings I put in the test config would be a bad idea -- if you want to fit a full quarterly dump, I'd recommend upping the memory for the test node to something like 16 gigs, so you'd modify the line: `"ES_JAVA_OPTS=-Xms512m -Xmx512m"` to be `"ES_JAVA_OPTS=-Xms16g -Xmx16g"`. If you're doing this on a laptop or some machine with limited memory, I would limit your data set to a lesser popular tld, such as `.biz` -- tld's like `.com` or `.org` are rather big.

**Comment by Mraoul** [comment]
I've released pydat 4.0 -- please use that and post any bugs you run into -- closing this issue
